api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java

```diff
@@ -991,4 +995,272 @@ public void checkIfPersonAttributeTypesAreLocked() {
+		if (preferred.getDeathDate() == null) {
+			preferred.setDeathDate(notPreferred.getDeathDate());
+		}
+		
+		if (preferred.getCauseOfDeath() != null) {
+			mergedData.setPriorCauseOfDeath(preferred.getCauseOfDeath().getUuid());
+		}
+		if (preferred.getCauseOfDeath() == null) {
+			preferred.setCauseOfDeath(notPreferred.getCauseOfDeath());
+		}
+	}
+	
+	private void mergeDateOfBirth(Person preferred, Person notPreferred, PersonMergeLogData mergedData) {
+		mergedData.setPriorDateOfBirth(preferred.getBirthdate());
+		mergedData.setPriorDateOfBirthEstimated(preferred.getBirthdateEstimated());
+		if (preferred.getBirthdate() == null || (preferred.getBirthdateEstimated() && !notPreferred.getBirthdateEstimated())) {
+			preferred.setBirthdate(notPreferred.getBirthdate());
+			preferred.setBirthdateEstimated(notPreferred.getBirthdateEstimated());
+		}
+	}
+	
+	private void mergePersonAttributes(Person preferred, Person notPreferred, PersonMergeLogData mergedData) {
+		// copy person attributes
+		for (PersonAttribute attr : notPreferred.getAttributes()) {
+			if (!attr.getVoided()) {
+				PersonAttribute tmpAttr = attr.copy();
+				tmpAttr.setPerson(null);
+				tmpAttr.setUuid(UUID.randomUUID().toString());
+				preferred.addAttribute(tmpAttr);
+				mergedData.addCreatedAttribute(tmpAttr.getUuid());
+			}
+		}
+	}
+	
+	private void mergeGenderInformation(Person preferred, Person notPreferred, PersonMergeLogData mergedData) {
+		// move all gender info
+		mergedData.setPriorGender(preferred.getGender());
+		if (!"M".equals(preferred.getGender()) && !"F".equals(preferred.getGender())) {
+			preferred.setGender(notPreferred.getGender());
+		}
+	}
+	
+	private void mergeNames(Person preferred, Person notPreferred, PersonMergeLogData mergedData) {
+		// move all names
+		// (must be done after all calls to services above so hbm doesn't try to save things prematurely (hacky)
+		for (PersonName newName : notPreferred.getNames()) {
+			boolean containsName = false;
+			for (PersonName currentName : preferred.getNames()) {
+				containsName = currentName.equalsContent(newName);
+				if (containsName) {
+					break;
+				}
+			}
+			if (!containsName) {
+				PersonName tmpName = constructTemporaryName(newName);
+				preferred.addName(tmpName);
+				mergedData.addCreatedName(tmpName.getUuid());
+				log.debug("Merging name " + newName.getGivenName() + " to " + preferred.getPersonId());
+			}
+		}
+	}
+	
+	private PersonName constructTemporaryName(PersonName newName) {
+		PersonName tmpName = PersonName.newInstance(newName);
+		tmpName.setPersonNameId(null);
+		tmpName.setVoided(false);
+		tmpName.setVoidedBy(null);
+		tmpName.setVoidReason(null);
+		// we don't want to change the preferred name of the preferred patient
+		tmpName.setPreferred(false);
+		tmpName.setUuid(UUID.randomUUID().toString());
+		return tmpName;
+	}
+	
+	private void mergeAddresses(Person preferred, Person notPreferred, PersonMergeLogData mergedData)
```

Comment:
This method could be break in to several methods. name of the method is called mergeAddresses but there are lot of things happening apart from merging addresses.

Reply:
I have changed this now...I was trying to use already written code in the patientserviceimpl class

Reply:
@reagan-meant I still cannot see your new changes

Reply:
@reagan-meant Also its not a good practice just to copy the code from patientserviceimpl. Better to move it to a common place and use it in Patientserviceimpl and PersonServiceImpl to avoid code duplication

Reply:
The original idea was to put all the merges in personservice class ..however when I tried it I realised I would still have code duplications since the three classes are only related but don't extend each other....this meant I would have similar method signatures that I have to overload for different classes....

Reply:
@reagan-meant How about moving the common methods to a util class and use it from there. for example if you create a class called mergeUtils and add mergeRelationships method there. Then use that method in both service classes ?

Reply:
@dilanthas I have failed to figure how to create a util class for two these different classes..